src/tracing/metric_event.h

```diff
@@ -0,0 +1,181 @@
+            uint32_t uint32_value;
+
+            Store(int32_t v) : int32_value(v) {}
+
+            Store(uint32_t v) : uint32_value(v) {}
+        };
+
+        Store store;
+        Type type;
+
+        Value(uint32_t value) : store(value), type(Type::kUInt32) {}
+
+        Value(int32_t value) : store(value), type(Type::kInt32) {}
+
+        Value(const ChipError & err) : store(err.AsInteger()), type(Type::kChipErrorCode) {}
+    };
+
+    MetricEvent(Type type, MetricKey key, int32_t value = 0) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, uint32_t value) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, const ChipError & error) : mType(type), mKey(key), mValue(error) {}
+
+    MetricEvent(Type type, MetricKey key, Value value) : mType(type), mKey(key), mValue(value) {}
+
+    MetricEvent(Type type, MetricKey key, int8_t value) : MetricEvent(type, key, int32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, uint8_t value) : MetricEvent(type, key, uint32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, int16_t value) : MetricEvent(type, key, int32_t(value)) {}
+
+    MetricEvent(Type type, MetricKey key, uint16_t value) : MetricEvent(type, key, uint32_t(value)) {}
+
+    Type type() const { return mType; }
+
+    MetricKey key() const { return mKey; }
+
+    Value value() const { return mValue; }
+
+    uint32_t ValueUInt32() const { return mValue.store.uint32_value; }
+
+    int32_t ValueInt32() const { return mValue.store.int32_value; }
+
+    uint32_t ValueErrorCode() const { return mValue.store.uint32_value; }
+
+private:
+    Type mType;
+    MetricKey mKey;
+    Value mValue;
+};
+
+namespace utils {
+
+/**
+ * Utility to emit an instant metric if the error is not a success.
+ */
+inline bool LogMetricIfError(MetricKey metricKey, const ::chip::ChipError & err)
+{
+    bool success = ::chip::ChipError::IsSuccess(err);
+    if (!success)
+    {
+        MATTER_LOG_METRIC(metricKey, err);
+    }
+    return success;
+}
+
+/**
+ * This utility class helps generate a Begin and End metric event within the scope of a block using RAII.
+ * This class is also meant to be used in expressions where ChipError object are typically used to capture
+ * error values.
+ */
+class ScopedMetricEvent
+{
+public:
+    ScopedMetricEvent(const ScopedMetricEvent &)             = default;
+    ScopedMetricEvent(ScopedMetricEvent &&)                  = default;
+    ScopedMetricEvent & operator=(const ScopedMetricEvent &) = default;
+    ScopedMetricEvent & operator=(ScopedMetricEvent &&)      = default;
+
+    ScopedMetricEvent(MetricKey key, const ChipError & error = CHIP_NO_ERROR) : mKey(key), mError(error)
+    {
+        MATTER_LOG_METRIC_BEGIN(mKey);
+    }
+
+    ~ScopedMetricEvent() { MATTER_LOG_METRIC_END(mKey, mError); }
+
+    operator ChipError() const { return mError; }
+
+    ScopedMetricEvent & operator=(const ChipError & err)
```

Comment:
This feels off ... an assignment is generally "take the value" however a scopedmetricevent is not an error, but rather contains an error however also contains a key or similar. Maybe `SetError` would be better than assignment operator.

Reply:
Maybe this is a name issue ... this seems to be a `TracedError` or `ScopedError` or something like that. It is not a "any type of metric" event. Lets rename.

Reply:
I can make this not to used as ChipError